File: test/family_link_contract_test.js
const test = require("node:test");
const assert = require("node:assert/strict");
const fs = require("node:fs");
const path = require("node:path");

const {
  FamilyBondPolicy,
  buildEncouragementPayload,
  buildFamilyLinkPayload,
  buildSharedGoalPayload,
} = require("../web_dashboard/assets/family_link_contract.js");

const cases = JSON.parse(
  fs.readFileSync(
    path.join(__dirname, "fixtures/family_link_contract_cases.json"),
    "utf8",
  ),
);

for (const testCase of cases) {
  test(testCase.name, () => {
    const build = () =>
      buildFamilyLinkPayload({
        senderId: testCase.senderId,
        senderRole: testCase.senderRole,
        receiverId: testCase.receiverId,
        receiverRole: testCase.receiverRole,
        now: "2026-07-27T00:00:00.000Z",
      });

    if (!testCase.valid) {
      assert.throws(build, /guardian and one child/);
      return;
    }

    const payload = build();
    assert.equal(payload.guardianId, testCase.guardianId);
    assert.equal(payload.childId, testCase.childId);
    assert.deepEqual(payload.participantIds.sort(), [
      testCase.childId,
      testCase.guardianId,
    ].sort());
    assert.equal(payload.status, "active");
    assert.equal(payload.consentScopes.summary, false);
    assert.equal(payload.consentScopes.healthTrends, false);
  });
}

test("encouragement targets the child without personal rewards", () => {
  const payload = buildEncouragementPayload({
    guardianId: "guardian-1",
    childId: "child-1",
    title: "今天也辛苦了",
    message: "先休息也沒關係",
    now: "2026-07-27T12:00:00.000Z",
  });

  assert.equal(payload.senderId, "guardian-1");
  assert.equal(payload.recipientId, "child-1");
  assert.equal(payload.status, "sent");
  assert.equal("disciplineCoins" in payload, false);
  assert.equal("avatarExperience" in payload, false);
});

test("shared goals need a child decision", () => {
  const payload = buildSharedGoalPayload({
    guardianId: "guardian-1",
    childId: "child-1",
    title: "每天專注 30 分鐘",
    message: "我們一起慢慢建立節奏",
    now: "2026-07-27T12:00:00.000Z",
  });

  assert.equal(payload.status, "proposed");
  assert.equal(payload.proposedBy, "guardian-1");
  assert.equal(payload.decisionBy, "child-1");
});

test("bond growth rewards two-way interaction", () => {
  assert.equal(FamilyBondPolicy.pointsFor("acknowledgement"), 3);
  assert.equal(FamilyBondPolicy.pointsFor("goalCompleted"), 10);
  assert.equal(FamilyBondPolicy.levelForXp(0), 1);
  assert.equal(FamilyBondPolicy.levelForXp(12), 2);
  assert.equal(FamilyBondPolicy.levelForXp(30), 3);
});

File: test/family_link_contract_test.dart
import 'dart:convert';
import 'dart:io';

import 'package:flutter_test/flutter_test.dart';
import 'package:nudge/models/family_link_contract.dart';

void main() {
  final cases =
      (jsonDecode(
                File(
                  'test/fixtures/family_link_contract_cases.json',
                ).readAsStringSync(),
              )
              as List)
          .cast<Map<String, dynamic>>();

  group('FamilyLinkContract', () {
    for (final testCase in cases) {
      test(testCase['name'] as String, () {
        FamilyLinkContract build() => FamilyLinkContract.fromAcceptedRequest(
          linkId: 'link-1',
          senderId: testCase['senderId'] as String,
          senderRole: testCase['senderRole'] as String,
          receiverId: testCase['receiverId'] as String,
          receiverRole: testCase['receiverRole'] as String,
          now: DateTime.utc(2026, 7, 27),
        );

        if (testCase['valid'] == false) {
          expect(build, throwsArgumentError);
          return;
        }

        final contract = build();
        expect(contract.guardianId, testCase['guardianId']);
        expect(contract.childId, testCase['childId']);
        expect(contract.participantIds, {
          testCase['guardianId'],
          testCase['childId'],
        });
        expect(contract.status, FamilyLinkStatus.active);
        expect(contract.consent.summary, isFalse);
        expect(contract.consent.healthTrends, isFalse);
      });
    }

    test(
      'encouragement targets the child and never mints personal rewards',
      () {
        final payload = FamilyLinkContract.buildEncouragementPayload(
          guardianId: 'guardian-1',
          childId: 'child-1',
          title: '今天也辛苦了',
          message: '先休息也沒關係',
          now: DateTime.utc(2026, 7, 27, 12),
        );

        expect(payload['senderId'], 'guardian-1');
        expect(payload['recipientId'], 'child-1');
        expect(payload['status'], 'sent');
        expect(payload.containsKey('disciplineCoins'), isFalse);
        expect(payload.containsKey('avatarExperience'), isFalse);
      },
    );

    test('shared goals begin proposed and need the child decision', () {
      final payload = FamilyLinkContract.buildSharedGoalPayload(
        guardianId: 'guardian-1',
        childId: 'child-1',
        title: '每天專注 30 分鐘',
        message: '我們一起慢慢建立節奏',
        now: DateTime.utc(2026, 7, 27, 12),
      );

      expect(payload['status'], 'proposed');
      expect(payload['proposedBy'], 'guardian-1');
      expect(payload['decisionBy'], 'child-1');
    });

    test('bond growth rewards two-way interaction, not message volume', () {
      expect(FamilyBondPolicy.pointsFor(FamilyBondEvent.acknowledgement), 3);
      expect(FamilyBondPolicy.pointsFor(FamilyBondEvent.goalCompleted), 10);
      expect(FamilyBondPolicy.levelForXp(0), 1);
      expect(FamilyBondPolicy.levelForXp(12), 2);
      expect(FamilyBondPolicy.levelForXp(30), 3);
    });
  });
}
